### PR TITLE
External IPs support

### PR DESCRIPTION
**Pull request by ArtfulCoder**

### CODE REVIEW

**Comment by ArtfulCoder** [comment]
@thockin This is work in progress. Added you on it so that you can see the changes in progress.

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -73,6 +74,8 @@ func (s *ProxyServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.ResourceContainer, "resource-container", s.ResourceContainer, "Absolute name of the resource-only container to create and run the Kube-proxy in (Default: /kube-proxy).")
 	fs.StringVar(&s.Kubeconfig, "kubeconfig", s.Kubeconfig, "Path to kubeconfig file with authorization information (the master location is set by the master flag).")
 	fs.Var(&s.PortRange, "proxy-port-range", "Range of host ports (beginPort-endPort, inclusive) that may be consumed in order to proxy service traffic. If unspecified (0-0) then ports will be randomly chosen.")
+	fs.Var(&s.PortRange, "proxy-port-range", "Range of host ports (beginPort-endPort, inclusive) that may be consumed in order to proxy service traffic. If unspecified (0-0) then ports will be randomly chosen.")
+	fs.IPNetVar(&s.ContainerCidr, "container-cidr", net.IPNet{}, "Cidr Range from which container IPs are allocated. If unspecified, external IPs will not be respected in the Kubernetes Service Spec.")
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: s/Cidr/CIDR/

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1088,7 +1088,7 @@ func ValidateService(service *api.Service) errs.ValidationErrorList {
 		}
 	}
 
-	for _, ip := range service.Spec.DeprecatedPublicIPs {
+	for _, ip := range service.Spec.ExternalIPs {
 		if ip == "0.0.0.0" {
 			allErrs = append(allErrs, errs.NewFieldInvalid("spec.publicIPs", ip, "is not an IP address"))
 		} else if util.IsValidIPv4(ip) && net.ParseIP(ip).IsLoopback() {
```
Comment (thockin) [inline pkg/api/validation/validation.go]: Something else we should think about - can any user create an ExternalIP that is the same as the node's IP on any port?  What happens if a user tries to claim 22?  I guess that can be fixed by my other PR to hold ports open - we should try to converge those to land at roughly the same time.

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: I have spent all evening thinking about this and I am not sure what is best.  Kubelet doesn't salt this in except on the master.  Maybe we should just do like kubelet - figure out the node name and ask the master for the node's CIDR.

If you're clever you can capture it as a library, maybe?

@dchen1107 for consult

Comment (thockin) [inline cmd/kube-proxy/app/server.go]: ohh, an idea.  I recalled reading something, and I googled it..

try `-m physdev ! --physdev-is-in` - this should only match packets from a bridge - some basic testing seems to work

File: pkg/proxy/userspace/proxier.go [hunk 3]
```diff
@@ -530,14 +538,19 @@ func (proxier *Proxier) closeOnePortal(portal portal, protocol api.Protocol, pro
...
+	}
+	args := proxier.iptablesContainerPortalArgs(portal.ip, excludedSource, portal.port, protocol, proxyIP, proxyPort, name)
 	if err := proxier.iptables.DeleteRule(iptables.TableNAT, iptablesContainerPortalChain, args...); err != nil {
 		glog.Errorf("Failed to delete iptables %s rule for service %q", iptablesContainerPortalChain, name)
 		el = append(el, err)
 	}
 
 	// Handle traffic from the host.
-	args = proxier.iptablesHostPortalArgs(portal.ip, portal.port, protocol, proxyIP, proxyPort, name)
+	args = proxier.iptablesHostPortalArgs(portal.ip, portal.isExternal, portal.port, protocol, proxyIP, proxyPort, name)
```
Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: We don't need to add external IPs to the host chain

File: pkg/proxy/userspace/proxier.go [hunk 4]
```diff
@@ -35,8 +35,9 @@ import (
 )
 
 type portal struct {
-	ip   net.IP
-	port int
+	ip         net.IP
+	port       int
+	isExternal bool
```
Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: don't stick this in the portal struct - just pass it as an extra param to openOnePortal

**Comment by thockin** [comment]
@brendandburns @bgrant0607 This PR basically brings back publicIPs under a different name with a clever twist - packets from the local host will not trigger the NAT.  This means the 8.8.8.8 style attack is neutralized.

I prefered the name "external" to "public" both because it seems more accurate and because it is a different semantic.  I could be argued down from this.

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (bgrant0607) [inline cmd/kube-proxy/app/server.go]: What about OVS?

Comment (bgrant0607) [inline cmd/kube-proxy/app/server.go]: BTW, yes, acronyms should have all the same case, so CIDR.

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/api/validation/validation.go]: Let's say no.

**Comment by bgrant0607** [comment]
Re. external vs. public: external is preferable. We don't use "public" anywhere else in the API, and we even described "publicIPs" using the word "external".

What I always found confusing about publicIPs was how it interacted with createExternalLoadBalancer. How does this interact with different values of Type?

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: awesome suggestion!
will try it out

Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: s/Cidr/CIDR - done. will be in next commit / update.

File: pkg/proxy/userspace/proxier.go (hunk 4 above)
Comment (ArtfulCoder) [inline pkg/proxy/userspace/proxier.go]: per offline conversation with Tim, leaving it as is. This code will hopefully be deleted/deprecated once we have proxying in kernel space.

**Comment by eparis** [comment]
@rajatchopra @dcbw @danwinship @pravisankar

**Comment by ArtfulCoder** [comment]
I believe I have addressed the code review comments..have pushed my changes.
I am still working on it..

**Comment by ArtfulCoder** [comment]
@thockin  
1. "-m physdev ! --physdev-is-in" matches traffic from docker0 bridge.
2. "-m physdev ! --physdev-is-in" does not match traffic coming with dstIP == forwardedIP

I will see if the physdev module has other matchers, but it seems that we have to go back to the original approach...

File: cmd/kube-proxy/app/server.go [hunk 5]
```diff
@@ -98,6 +98,7 @@ func (s *ProxyServer) Run(_ []string) error {
 		protocol = iptables.ProtocolIpv6
 	}
 	loadBalancer := userspace.NewLoadBalancerRR()
+
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: This is a pointless hunk.

File: pkg/kubectl/cmd/expose.go [hunk 6]
```diff
@@ -67,7 +67,7 @@ func NewCmdExposeService(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().Bool("dry-run", false, "If true, only print the object that would be sent, without creating it.")
 	cmd.Flags().String("container-port", "", "Synonym for --target-port")
 	cmd.Flags().String("target-port", "", "Name or number for the port on the container that the service should direct traffic to. Optional.")
-	cmd.Flags().String("public-ip", "", "Name of a public IP address to set for the service. The service will be assigned this IP in addition to its generated service IP.")
+	cmd.Flags().String("external-ips", "", "External IP addresses(comma-separated) to set for the service. The service will be respond to these IPs in addition to its generated service IP.")
```
Comment (thockin) [inline pkg/kubectl/cmd/expose.go]: "will be respond" is nonsense

File: pkg/proxy/userspace/proxier.go (hunk 4 above)
Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: In the iptables form (once my PR goes in) this struct is gone.

File: pkg/proxy/userspace/proxier.go [hunk 7]
```diff
@@ -687,12 +696,16 @@ func iptablesCommonPortalArgs(destIP net.IP, destPort int, protocol api.Protocol
 		args = append(args, "-d", fmt.Sprintf("%s/32", destIP.String()))
 	}
 
+	if excludeContainerSourceTraffic {
+		args = append(args, "-m", "physdev", "--physdev-is-in")
```
Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: don't you want `! --physdev-is-in` ?  You want to exclude container traffic...

**Comment by thockin** [comment]
@ArtfulCoder 

I don't understand:

> 1. "-m physdev --physdev-is-in" matches traffic from docker0 bridge.
> 2. "-m physdev --physdev-is-in" does not match traffic coming with dstIP == forwardedIP

First, I think you want to negate, so `-m physdev ! --physdev-is-in` will only match traffic that DID NOT come from docker0.

I don't understand your #2 point.  Once you negate, you should still be receiving external traffic, right?  What am I missing?

Also, you need to rebase this onto iptables proxier, or at least make it compile.  I have #12764 pending - so it would be nice to rebase onto that, but if you finish before me, I'll deal with it.

File: pkg/proxy/userspace/proxier.go (hunk 7 above)
Comment (ArtfulCoder) [inline pkg/proxy/userspace/proxier.go]: I need "!" if use "-j DROP". (negative filter)
I did not need to use "!" because I send the packet to proxier on the physdev-is-in match (-j REDIRECT). (positive filter)

**Comment by ArtfulCoder** [comment]
@thockin 
For a moment, lets just see what the iptable matchers do:

Based on the manual setup on GCE node, I found that: 
"-m physdev --physdev-is-in" matches traffic from docker0 bridge. 
In essence, It does not help differentiate between traffic from dbr0 v/s eth0, which was our goal.

"-m physdev --physdev-is-in" does not match traffic coming with dstIP == forwardedIP
What does means is that, packets that arrive with a GCE Forwarded IP as the destination IP, are not considered to arrive on the interface. Maybe because it does not match the ip assigned to any interface ?

You mentioned that in your experiment, you had it working. 
But, i am not seeing it in my test...

File: pkg/proxy/userspace/proxier.go (hunk 7 above)
Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: This is not a filter rule at all, it's a nat rule.  You want to match not-bridged packets, which sends them to the proxier.  Bridged packets will go through the normal packet path.   I feel like I am missing something - how would this work at all as it is?

Comment (ArtfulCoder) [inline pkg/proxy/userspace/proxier.go]: Here is an example of what the code would do for external ip 104.154.64.226: 
-A KUBE-PORTALS-CONTAINER -d 104.154.64.226/32 -p tcp -m comment --comment "kube-system/mypod-svc:" -m tcp --dport 9091 -m physdev --physdev-is-in -j REDIRECT --t
o-ports 39236

I used the term "filter" in an overloaded way. It is a nat rule.

Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: Ahh, I see you have named a variable backwards.  This flag is called "exclude" but you pass false to it when you want to exclude.  Try it the other way.

You only care about whether it is a bridged packet IFF the caller said to exclude.

clusterIP: don't care about bridging
load-balancer: don't care about bridging
external IP: don't allow bridging

I think that handles what you want here.

Comment (ArtfulCoder) [inline pkg/proxy/userspace/proxier.go]: yes, will try it now

Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: > Here is an example of what the code would do for external ip 104.154.64.226:
> -A KUBE-PORTALS-CONTAINER -d 104.154.64.226/32 -p tcp -m comment --comment "kube-system/mypod-svc:" -m tcp --dport 9091 -m physdev --physdev-is-in -j REDIRECT --t
> o-ports 39236

Yeah, this is wrong.  You want `-d 104.154.64.226/32` and `!
--physdev-is-in` - think about it.  Traffic bound for the external IP
and NOT from a local container.

Also please leave a TODO that this will not work for situations that
do not involve a bridge (e.g. network plugins).

Comment (ArtfulCoder) [inline pkg/proxy/userspace/proxier.go]: yes, I am missing the !
My understanding of physdev-is-in was wrong.
I thought it was for non-virtual interface (hence the term physdev)
but it seems to be true for bridged/virtual interfaces.

**Comment by ArtfulCoder** [comment]
## Setup
We have a http container running on NodeA on the GCE Kubernetes cluster.
We have set it to listen on port 9091 for external vip 104.154.64.226

### iptables nat rule:
`-A KUBE-PORTALS-CONTAINER -d 104.154.64.226/32 -p tcp -m comment --comment "kube-system/mypod-svc:" -m tcp --dport 9091 -m physdev ! --physdev-is-in -j REDIRECT --to-ports 41496`

Assume that NodeA is in the target pool of the VIP
Assume that NodeB is not in the target pool of the VIP
Assume that ContainerA is on NodeA
Assume that ContainerB in on NodeB


## Curl results for : `curl http://104.1.54.64.226:9091`
Non-Google-Network | NodeA  | ContainerA | NodeB | ContainerB
------------ | ------------ | ------------- | ------------- | ------------- 
Success | Fail |  Fail | Success | Success

Details below:

## Curl from non-google network (it works)
`curl http://104.1.54.64.226:9091`

### Corresponding tcpdump listed below  ( we see the syn, syn-ack, ack packets, 3 packet lines below)

tcpdump -i any -n host 104.154.64.226
`tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 65535 bytes
20:18:30.680990 IP 216.239.45.81.36282 > 104.154.64.226.9091: Flags [S], seq 1085801095, win 65535, options [mss 1086,nop,wscale 4,nop,nop,TS val 1991279150 ecr 0,sackOK,eol], length 0`

`20:18:30.681036 IP 104.154.64.226.9091 > 216.239.45.81.36282: Flags [S.], seq 4261993958, ack 1085801096, win 28160, options [mss 1420,sackOK,TS val 45483539 ecr 1991279150,nop,wscale 7], length 0`

`20:18:30.734451 IP 216.239.45.81.36282 > 104.154.64.226.9091: Flags [.], ack 1, win 8256, options [nop,nop,TS val 1991279204 ecr 45483539], length 0`


## Curl from same node (doesnt work)
`curl http://104.154.64.226:9091`
`curl: (7) couldn't connect to host`

### Corresponding tcpdump  ( we see the syn, reset-ack packets, but not any syn-ack or ack packets)
`tcpdump -i any -n host 104.154.64.226`
`tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 65535 bytes`
`20:26:57.963321 IP 104.154.64.226.34544 > 104.154.64.226.9091: Flags [S], seq 4159492410, win 43690, options [mss 65495,sackOK,TS val 45610359 ecr 0,nop,wscale 7], length 0`
`20:26:57.963340 IP 104.154.64.226.9091 > 104.154.64.226.34544: Flags [R.], seq 0, ack 4159492411, win 0, length 0`


## Curl from any docker container on the same node. (doesnt work)
`docker exec -ti f6  curl http://104.154.64.226:9091
curl: (7) Failed to connect to 104.154.64.226 port 9091: Connection refused`


### Corresponding tcpdump  ( we see the syn, reset-ack packets, but not any syn-ack or ack packets)
`tcpdump -i any -n host 104.154.64.226` 
`tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 65535 bytes`
`20:22:35.800045 IP 10.244.2.7.50584 > 104.154.64.226.9091: Flags [S], seq 3990982479, win 28400, options [mss 1420,sackOK,TS val 45544818 ecr 0,nop,wscale 7], length 0`
`20:22:35.800045 IP 10.244.2.7.50584 > 104.154.64.226.9091: Flags [S], seq 3990982479, win 28400, options [mss 1420,sackOK,TS val 45544818 ecr 0,nop,wscale 7], length 0`
`20:22:35.800073 IP 104.154.64.226.9091 > 10.244.2.7.50584: Flags [R.], seq 0, ack 3990982480, win 0, length 0`
`20:22:35.800075 IP 104.154.64.226.9091 > 10.244.2.7.50584: Flags [R.], seq 0, ack 1, win 0, length 0`


## Curl from another node (or docker container on another node) that is not part of the vip's target pool
`curl http://104.154.64.226:9091`
It works.

### Corresponding tcpdump ( we see the syn, syn-ack, ack packets, 3 packet lines below)
`tcpdump -i any -n host 104.154.64.226` 
`tcpdump: verbose output suppressed, use -v or -vv for full protocol decode`
`listening on eth0, link-type EN10MB (Ethernet), capture size 65535 bytes`

`20:36:18.813744 IP 104.197.23.196.33357 > 104.154.64.226.9091: Flags [S], seq 1399958873, win 28400, options [mss 1420,sackOK,TS val 45752694 ecr 0,nop,wscale 7], length 0`

`20:36:18.813791 IP 104.154.64.226.9091 > 104.197.23.196.33357: Flags [S.], seq 2367011011, ack 1399958874, win 28160, options [mss 1420,sackOK,TS val 45750572 ecr 45752694,nop,wscale 7], length 0`

`20:36:18.814833 IP 104.197.23.196.33357 > 104.154.64.226.9091: Flags [.], ack 1, win 222, options [nop,nop,TS val 45752695 ecr 45750572], length 0`

**Comment by thockin** [comment]
I'm assuming this is because GCE adds the IP of any load-balancer that
targets a VM to that VM's local IP address tables.  Accessing that IP would
be like hitting the loopback interface, which is (of course) not going to
work because that whole model assumes VMs are the target, and not
containers.

TL;DR not surprised, and not super concerned about it.

Maybe we could add a second rule:

1: capture all traffic to LBIP which is from off-machine `-A PREROUTING ...
-m physdev ! --physdev-is-in ...`
2: capture all traffic to LBIP which is bound for a local address `A
{PREROUTING,OUTPUT} ... -m addrtype -dst-type LOCAL ...`

If the origin of traffic is off-machine, it will always hit #1, whether the
machine know the destination IP or not - someone arranged for that traffic
to arrive here.

If the origin of traffic is on-machine AND the IP is claimed by the
machine, it will hit #2.  I wonder if there's actually a need for
off-machine traffic that is NOT claimed as a local IP, but it seems
plausible.

So now your 1 rule becomes 3.  With excellent comments, of course.  Try it
out?

**Comment by ArtfulCoder** [comment]
excellent idea!
will try it out and see if it covers all cases and is safe.

File: pkg/kubectl/cmd/expose.go (hunk 6 above)
Comment (ArtfulCoder) [inline pkg/kubectl/cmd/expose.go]: fixed

File: pkg/api/validation/validation.go [hunk 8]
```diff
@@ -1765,6 +1765,23 @@ func validateEndpointAddress(address *api.EndpointAddress) errs.ValidationErrorL
 	return allErrs
 }
 
+func ValidateIpIsNotLinkLocalOrLoopback(ipAddress string) errs.ValidationErrorList {
```
Comment (thockin) [inline pkg/api/validation/validation.go]: This should be called from ValidateService() in this file, which currently tests DeprecatedPublicIPs for 0.0.0.0 and loopback

File: pkg/proxy/iptables/proxier.go [hunk 9]
```diff
@@ -564,10 +564,17 @@ func (proxier *Proxier) syncProxyRules() error {
 				"-d", fmt.Sprintf("%s/32", externalIP),
 				"--dport", fmt.Sprintf("%d", info.port),
 			}
-			// We have to SNAT packets from external IPs.
+			// We have to SNAT packets to external IPs.
 			writeLine(rulesLines, append(args,
 				"-j", "MARK", "--set-xmark", fmt.Sprintf("%s/0xffffffff", iptablesMasqueradeMark))...)
-			writeLine(rulesLines, append(args,
+
+			externalTrafficOnlyArgs := append(args,
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: Comment, something like

Allow traffic that is not from a bridge (i.e. not from a container) and not from a local process to be forwarded to the service.  This roughly translates to "all traffic from off-machine".  This is imperfect in the face of network plugins that might not use a bridge, but we can revisit that later.

File: pkg/proxy/iptables/proxier.go [hunk 10]
```diff
@@ -564,10 +564,17 @@ func (proxier *Proxier) syncProxyRules() error {
...
+				"-m", "physdev", "!", "--physdev-is-in",
+			    "-m", "addrtype", "!", "--src-type", "LOCAL")
+			writeLine(rulesLines, append(externalTrafficOnlyArgs,
+				"-j", string(svcChain))...)
+			dstLocalOnlyArgs := append(args,  "-m", "addrtype", "--dst-type", "LOCAL")
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: Comment, something like

Allow traffic bound for external IPs that happen to be recognized as local IPs to stay local.  This covers cases like GCE load-balancers which get added to the local routing table.

File: pkg/registry/service/rest.go [hunk 11]
```diff
@@ -115,6 +115,12 @@ func (rs *REST) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
 		}
 	}
 
+	for _, externalIP := range service.Spec.ExternalIPs {
```
Comment (thockin) [inline pkg/registry/service/rest.go]: as per other comment,. this should not be needed

**Comment by ArtfulCoder** [comment]
PTAL

File: pkg/api/validation/validation.go [hunk 12]
```diff
@@ -1090,12 +1090,13 @@ func ValidateService(service *api.Service) errs.ValidationErrorList {
...
 	}
 
-	for _, ip := range service.Spec.DeprecatedPublicIPs {
+	for _, ip := range service.Spec.ExternalIPs {
 		if ip == "0.0.0.0" {
 			allErrs = append(allErrs, errs.NewFieldInvalid("spec.publicIPs", ip, "is not an IP address"))
 		} else if util.IsValidIPv4(ip) && net.ParseIP(ip).IsLoopback() {
 			allErrs = append(allErrs, errs.NewFieldInvalid("spec.publicIPs", ip, "publicIP cannot be a loopback"))
 		}
+		allErrs = append(allErrs, ValidateIpIsNotLinkLocalOrLoopback(ip)...)
```
Comment (thockin) [inline pkg/api/validation/validation.go]: just 3 lines above this we check for loopback - that is redundant with this.  de-dup, please.

File: pkg/api/validation/validation.go [hunk 13]
```diff
@@ -1090,12 +1090,13 @@ func ValidateService(service *api.Service) errs.ValidationErrorList {
 		}
 	}
 
-	for _, ip := range service.Spec.DeprecatedPublicIPs {
+	for _, ip := range service.Spec.ExternalIPs {
 		if ip == "0.0.0.0" {
 			allErrs = append(allErrs, errs.NewFieldInvalid("spec.publicIPs", ip, "is not an IP address"))
```
Comment (thockin) [inline pkg/api/validation/validation.go]: s/publicIPs/externalIPs/g - search whole file, please

File: pkg/kubectl/cmd/expose.go [hunk 14]
```diff
@@ -70,7 +70,7 @@ func NewCmdExposeService(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().Bool("dry-run", false, "If true, only print the object that would be sent, without creating it.")
 	cmd.Flags().String("container-port", "", "Synonym for --target-port")
 	cmd.Flags().String("target-port", "", "Name or number for the port on the container that the service should direct traffic to. Optional.")
-	cmd.Flags().String("public-ip", "", "Name of a public IP address to set for the service. The service will be assigned this IP in addition to its generated service IP.")
+	cmd.Flags().String("external-ips", "", "External IP addresses(comma-separated) to set for the service. The service can be accessed by these IPs in addition to its generated service IP.")
```
Comment (thockin) [inline pkg/kubectl/cmd/expose.go]: this should use a StringSlice flag rather than manual parsing

https://godoc.org/github.com/spf13/pflag#StringSlice

Comment (thockin) [inline pkg/kubectl/cmd/expose.go]: @eparis it doesn't look like pflags.StringSlice and Cobra are well integrated - interesting to you?

Comment (thockin) [inline pkg/kubectl/cmd/expose.go]: @ArtfulCoder just make it singular for now

File: hack/verify-flags/known-flags.txt [hunk 15]
```diff
@@ -77,6 +77,7 @@ executor-suicide-timeout
 experimental-keystone-url
 experimental-prefix
 external-hostname
+external-ips
```
Comment (thockin) [inline hack/verify-flags/known-flags.txt]: need to remove old public-ip flag?

**Comment by thockin** [comment]
LG except flags

**Comment by ArtfulCoder** [comment]
fixed flags. PTAL

**Comment by thockin** [comment]
lgtm

**Comment by saad-ali** [comment]
Rebase needed

**Comment by thockin** [comment]
rebased - @k8s-oncall please merge when safe

**Comment by zmerlynn** [comment]
Shippable flaked; kicked.

**Comment by saad-ali** [comment]
Will merge on green

**Comment by davidopp** [comment]
Should you add documentation about this to
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/services.md
? (or somewhere?) Sorry if you did and I missed it.

**Comment by thockin** [comment]
Good point!  I suck.  Abhi, can you write a bit about this for the user
guide, and maybe a "gory details" section at the bottom covering the
special-cases

**Comment by ArtfulCoder** [comment]
Will do.

**Comment by bgrant0607** [comment]
+1 to documentation

**Comment by BugRoger** [comment]
Excuse the necro of this issue but it seems to relate best to my question.

We are running a bare-metal cluster in an environment where we don't have much else beside Kubernetes. Especially there is no external load-balancing device that could be remote controlled. Naturally that begs the question how to expose services. 

Our current approach is to have a subnet that we use for external IPs. Despite it being "clumsy" we do the bookkeeping of used IPs with a human workflow and the networking with some BGP magic. So far so good.

Now we ran into the trouble that our application (cough, OpenStack) is not really "container-aware". It brings it's own service discovery catalog that even though it has the notion of public and internal service endpoints does not make use of it consistently. We found that it often defaults to the public endpoints even though an internal one would be more appropriate. To make matters worse, the system can't be completely containerized and there components running outside of Kubernetes which can't use the service ips.

Now the trouble is that with the scenario above, we sometimes run into the problem that a container can't reach the external ip which is exposed on the same node it is running on (the test case from above "Curl from any docker container on the same node. (doesnt work)"). It works though from any other node in the cluster.

The crude workaround we are thinking about now is to bind all IPs of the external subnet to each node. Meh.

TL;DR:

  * Is it intended that containers should be able to reach the external IP? 
  * Is this a bug or a feature that they can't do so reliably? :smile:

Before we run off with a workaround that relies on an unintended side-effect, would you (@thockin, @ArtfulCoder) mind clarifying the intentions here. /bow

**Comment by smarterclayton** [comment]
I thought there were security implications if the container can reach the
external IP (because any service can be hijacked by another service that
sets its externalIP = otherService.clusterIP).

**Comment by SchwarzM** [comment]
Due to the nature of our networking (bgp) we need to define a node of the cluster that gets the traffic for the particular external ip. This implies that the rest of the cluster will gain a physical hop to the external ip.
Rendering the external ip to what used to be a public ip.
The node itself will then not be able to reach the external ip due to it coming from a bridged interface and not having the ip bound locally.
The obvious solution would be to bind all the external service ip's locally (e.g. to a dummy network).

It seems a strange behaviour thus we would like to know the reasoning behind the external ip design to align our behaviour with the reasons and not the particular implementation (the iptables rules as they are).

Kind regards.
